### PR TITLE
[SPARK-35112][SQL] Support Cast string to day-second interval

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
Support Cast string to day-seconds interval

### Why are the changes needed?
Users can cast day-second interval string to DayTimeIntervalType.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added UT

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
FYI @MaxGekk

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 1]
```diff
@@ -123,6 +123,24 @@ object IntervalUtils {
...
   }
 
+  def safeFromDayTimeString(input: UTF8String): Option[Long] = {
+    try {
+      if (input == null || input.toString == null) {
+        throw new IllegalArgumentException("Interval day-second string must be not null")
+      } else {
+        val regex = "INTERVAL '([-|+]?[0-9]+ [0-9:]+)' DAY TO SECOND".r
+        // scalastyle:off caselocale .toLowerCase
+        val intervalString = input.trimAll().toUpperCase.toString
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:133]: I think you can make the regexp case insensitive.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 2]
```diff
@@ -123,6 +123,24 @@ object IntervalUtils {
...
+        // scalastyle:on
+        val interval = regex.findFirstMatchIn(intervalString)
+          .map(_.group(1)).getOrElse(intervalString)
+        Some(fromDayTimeString(interval).microseconds)
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:137]: `fromDayTimeString()` sets `days` too.

**Review by MaxGekk — CHANGES_REQUESTED** [review]
Take into account `days` in safeFromDayTimeString()

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala (hunk 2 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:137]: Please, look at https://github.com/apache/spark/blob/e08c40fa3f7054bdc713873c99d3aaf0014d9314/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala#L2326

**Comment by AngersZhuuuu** [comment]
@MaxGekk I meet a case 

IntervalUtils.fromDayTimeString("2:03:04") throw exception
```
[info]   Cause: java.lang.IllegalArgumentException: requirement failed: Interval string must match day-time format of '^(?<sign>[+|-])?(?<day>\d+) (?<hour>\d{1,2}):(?<minute>\d{1,2}):(?<second>(\d{1,2})(\.(\d{1,9}))?)$': 2:03:04, set spark.sql.legacy.fromDayTimeString.enabled to true to restore the behavior before Spark 3.0.
```
It's a But right? since `IntervalUtils.fromDayTimeString`'s comment is 
```
  /**
   * Parse dayTime string in form: [-]d HH:mm:ss.nnnnnnnnn and [-]HH:mm:ss.nnnnnnnnn
   *
   * adapted from HiveIntervalDayTime.valueOf
   */
  def fromDayTimeString(s: String): CalendarInterval = {
    fromDayTimeString(s, DAY, SECOND)
  }
```

**Comment by AngersZhuuuu** [comment]
Current test 
```
Incorrect evaluation (codegen off): cast(cast(INTERVAL '106751991 04:00:54.775807' DAY TO SECOND as string) as day-time interval), actual: 14454775807, expected: 9223372036854775807 (ExpressionEvalHelper.scala:209)
```

**Comment by MaxGekk** [comment]
@AngersZhuuuu Is it still WIP? Please, ping me when it is ready.

**Comment by AngersZhuuuu** [comment]
@MaxGekk Can be reviewed now,  but currently this method does not seem to be very efficient，Maybe should I add a new method to calculate  calculate total microseconds separately?

**Comment by AngersZhuuuu** [comment]
@MaxGekk  I think it's ready for review now, but I am confused how can I use 
`dayTimePattern(DAY -> SECOND)` in match case ?

When I write 
```
val regex = dayTimePattern(DAY-> SECOND)

intervalStr match {
    case dayTimePattern(sign, day, hour, minute, secondAndMicro)
.......
}
```
It can't match, do you know how to match this? If it can be done, we can simplify more code.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala [hunk 3]
```diff
@@ -1775,6 +1775,48 @@ class CastSuite extends CastSuiteBase {
     }
   }
 
+  test("SPARK-35112: Cast string to day-time interval") {
+    checkEvaluation(cast(Literal.create("0 0:0:0"), DayTimeIntervalType), 0L)
+    checkEvaluation(cast(Literal.create("INTERVAL '0 0:0:0' DAY TO SECOND"),
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:1780]: Could you check `INTERVAL` in lower case + spaces:
```suggestion
    checkEvaluation(cast(Literal.create(" interval '0 0:0:0' Day TO second   "),
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 4]
```diff
@@ -150,6 +150,57 @@ object IntervalUtils {
     }
   }
 
+  private val daySecondStringPattern = ("(?i)^(INTERVAL\\s+)([+|-])?(')" +
+    "([+|-])?(\\d+) (\\d{1,2}):(\\d{1,2}):(\\d{1,2})(\\.\\d{1,9})?(')(\\s+DAY\\s+TO\\s+SECOND)$").r
+  private val daySecondPattern = "^([+|-])?(\\d+) (\\d{1,2}):(\\d{1,2}):(\\d{1,2})(\\.\\d{1,9})?$".r
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:155]: Let's extract the common pattern:
```suggestion
  private val unquotedDaySecondPattern =
    "([+|-])?(\\d+) (\\d{1,2}):(\\d{1,2}):(\\d{1,2})(\\.\\d{1,9})?"
  private val quotedDaySecondPattern = (s"^$unquotedDaySecondPattern$$").r
  private val daySecondLiteralPattern =
    (s"(?i)^(INTERVAL\\s+)([+|-])?(')$unquotedDaySecondPattern(')(\\s+DAY\\s+TO\\s+SECOND)$$").r
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 5]
```diff
@@ -150,6 +150,57 @@ object IntervalUtils {
...
+  def castStringToDTInterval(input: UTF8String): Long = {
+    def secondAndMicro(second: String, micro: String): String = {
+      if (micro != null) {
+        s"$second$micro"
+      } else {
+        second
+      }
+    }
+    val intervalStr = input.trimAll().toString
+    intervalStr match {
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:166]: Just inline it:
```suggestion
    input.trimAll().toString match {
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 6]
```diff
@@ -150,6 +150,57 @@ object IntervalUtils {
...
+      case daySecondPattern("-", day, hour, minute, second, micro) =>
+        toDTInterval(day, hour, minute, secondAndMicro(second, micro), -1)
+      case daySecondPattern(_, day, hour, minute, second, micro) =>
+        toDTInterval(day, hour, minute, secondAndMicro(second, micro), 1)
+      case daySecondStringPattern(
+      _, firstSign, _, secondSign, day, hour, minute, second, micro, _, _) =>
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:172]: You don't need to put quotes to groups:
```suggestion
      case daySecondLiteralPattern(firstSign, _, secondSign, day, hour, minute, second, micro) =>
```
if you change the pattern to:
```
  private val daySecondLiteralPattern =
    (s"(?i)^(INTERVAL\\s+)([+|-])?\\'$unquotedDaySecondPattern\\'\\s+DAY\\s+TO\\s+SECOND$$").r
```

**Review by MaxGekk — COMMENTED** [review]
LGTM in general except a few comments.

**Review by MaxGekk — APPROVED** [review]
LGTM, waiting for test results.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala (hunk 3 above)
Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:1780]: @AngersZhuuuu Your force push reverted the changes I approved. Please, revert them back.

**Review by MaxGekk — CHANGES_REQUESTED** [review]
Please, address comments.

**Comment by AngersZhuuuu** [comment]
> Please, address comments.

Sorry for my late, since yesterday's run failed so I revert and recheck, but some thing urgent break my work.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala (hunk 3 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala]: > @AngersZhuuuu Your force push reverted the changes I approved. Please, revert them back.

Updated

**Review by MaxGekk — APPROVED** [review]
+1, LGTM. Merging to master.
Thank you, @AngersZhuuuu .

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 7]
```diff
@@ -150,6 +150,58 @@ object IntervalUtils {
     }
   }
 
+  private val unquotedDaySecondPattern =
+    "([+|-])?(\\d+) (\\d{1,2}):(\\d{1,2}):(\\d{1,2})(\\.\\d{1,9})?"
+  private val quotedDaySecondPattern = (s"^$unquotedDaySecondPattern$$").r
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:155]: `unquotedDaySecondRegex`?

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:155]: > `unquotedDaySecondRegex`?

Change this in https://github.com/apache/spark/pull/32444

**Pull request closed without merging**